src/promptflow/promptflow/contracts/types.py

```diff
@@ -29,13 +30,23 @@ class FilePath(str):
     pass
 
 
+@dataclass
+class AssistantTool:
+    name: str
+    openai_definition: dict
+    func: Callable
+
+
 @dataclass
 class AssistantDefinition:
     """This class is used to define an assistant definition."""
 
     model: str
     instructions: str
-    tools: list
+    # The raw tool definition in json string
+    tools: List
+    # The resolved tool definition by name. This property is addon and will not be part of pickling.
+    assistant_tools: Dict[str, AssistantTool] = field(default_factory=dict)
```

Comment:
I don't think it is a contract level concept, it is just about implementation, please try another way.

Reply:
Per offline discussion, we use implicit attribute to contain the resolved part